python/ray/data/_internal/execution/operators/zip_operator.py

```diff
@@ -236,11 +237,32 @@ def _zip_one_block(
         # Swap blocks if ordering was inverted during block alignment splitting.
         block, other_block = other_block, block
     # Zip block and other blocks.
-    result = BlockAccessor.for_block(block).zip(other_block)
+    result = _try_zip_with_pyarrow_fallback(block, other_block)
     br = BlockAccessor.for_block(result)
     return result, br.get_metadata(input_files=[], exec_stats=stats.build())
 
 
 def _get_num_rows_and_bytes(block: Block) -> Tuple[int, int]:
     block = BlockAccessor.for_block(block)
     return block.num_rows(), block.size_bytes()
+
+
+def _try_zip_with_pyarrow_fallback(block: Block, other_block: Block) -> Block:
```

Comment:
Rather than introducing a wrapping function, would it make sense to move this try-catch logic to `Block.zip()`? That way, you wouldn't need to call `_try_zip_with_pyarrow_fallback` whenever you call `zip()`

Reply:
Moved all the try logic to here after addressing these comments:
https://github.com/ray-project/ray/pull/39817#discussion_r1336288548
https://github.com/ray-project/ray/pull/39817#discussion_r1336288340

Reply:
one alternative approach could be to move this logic to `TableBlockAccessor.zip()`, and replace the logic for raising a `ValueError` if the BlockAccessor types don't match with the conversion logic implemented in `_try_zip_with_pyarrow_fallback`.

Reply:
Sure. Happy to make that change if we believe that is where the arrow fallback logic should be.

Reply:
Here is one way to move everything into table_block.py
```diff
-    result = BlockAccessor.for_block(block).zip(other_block)
+    result = _try_zip_with_pyarrow_fallback(block, other_block)
     br = BlockAccessor.for_block(result)
     return result, br.get_metadata(input_files=[], exec_stats=stats.build())
 
@@ -244,3 +244,21 @@ def _zip_one_block(
 def _get_num_rows_and_bytes(block: Block) -> Tuple[int, int]:
     block = BlockAccessor.for_block(block)
     return block.num_rows(), block.size_bytes()
+
+
+def _try_zip_with_pyarrow_fallback(block: Block, other_block: Block) -> Block:
+    """Try zipping blocks together. If zip fails on ValueError exception due to
+    mismatched block formats we convert `block` or `other_block` to arrow format
+    and zip again.
+    """
+    try:
+        return BlockAccessor.for_block(block).zip(other_block)
+    except ValueError:
+        import pyarrow
+
+        if isinstance(other_block, pyarrow.Table):
+            block = BlockAccessor.for_block(block).to_arrow()
+            return BlockAccessor.for_block(block).zip(other_block)
+        else:
+            other_block = BlockAccessor.for_block(other_block).to_arrow()
+            return BlockAccessor.for_block(block).zip(other_block)
```

Reply:
Here is a version with everything in table_block.py:
```diff 
# 4efe48349fda3b8e66fe13bf4cb65cd514f8a5f1
+++ b/python/ray/data/_internal/table_block.py
@@ -206,15 +206,28 @@ class TableBlockAccessor(BlockAccessor):
     def zip(self, other: "Block") -> "Block":
         acc = BlockAccessor.for_block(other)
         if not isinstance(acc, type(self)):
-            raise ValueError(
-                "Cannot zip {} with block of type {}".format(type(self), type(other))
-            )
+            from pyarrow import Table
+
+            if isinstance(self._table, Table):
+                other = acc.to_arrow()
+                return self._zip(BlockAccessor.for_block(other))
+            elif isinstance(other, Table):
+                block = self.to_arrow()
+                return BlockAccessor.for_block(block).zip(other)
+            else:
+                raise ValueError(
+                    "Cannot zip {} with block of type {}".format(
+                        type(self), type(other)
+                    )
+                )
+
         if acc.num_rows() != self.num_rows():
             raise ValueError(
                 "Cannot zip self (length {}) with block of length {}".format(
                     self.num_rows(), acc.num_rows()
                 )
             )
+
         return self._zip(acc)
 ```

Reply:
I committed the hybrid.